tracks/julia/exercises/rotational-cipher/mentoring.md

```diff
@@ -0,0 +1,80 @@
+### Mentor guidance
+
+We want them to realise that they can define two methods for `rotate`.
+
+The metaprogramming is quite challenging for lots of students. Consider offering to just show them the answer: "The metaprogramming is a bit hard to understand at first, would you like me to show you that part of the answer?"
+
+
+### Example solution
+
+```julia
+"""
+    rotate(n, c::AbstractChar)
+
+Rotate `c` `n` places if it is an ASCII letter, else return it
+
+"""
+function rotate(n, c::AbstractChar)
+    if c in 'a':'z'
+        'a' + (c - 'a' + n) % 26
+    elseif c in 'A':'Z'
+        'A' + (c - 'A' + n) % 26
+    else
+        c
+    end
+end
+
+"""
+    rotate(n, itr)
+
+Rotate each ASCII alphabetic character in iterable `n` places
+
+"""
+rotate(n, itr) = map(c -> rotate(n, c), itr)
+```
+
+Metaprogramming bit:
+
+Just R13:
+
+```
+macro R13_str(s)
+    :(rotate(13, $s))
+end
+```
+
+All of the macros done with a for loop:
+
+```
+for n in 0:26
+    @eval macro $(Symbol(:R, n, :_str))(s)
+        :(rotate($$n, $s))
+    end
+end
+```
+
+
+### Common suggestions
+
+(Formatted so you can copy and paste easily)
+
+- The metaprogramming is a bit hard to understand at first, would you like me to show you that part of the answer?
+- It's often useful to break a problem down into smaller parts. Perhaps you'd like to start with this?
+
+```
+"Explain what this method does here"
+function rotate(n, c::Char)
+    ...
+end
+
+"Explain what this method does here"
+function rotate(n, str)
+    ...
+end
+```
+
+
+### Talking points
+
+- You can make this a fair bit faster if you trust the string is ASCII and use codeunits() and some unsigned integer arithmetic. Unsigned int literals look like 0xF1 or 0b11110001
```

Comment:
I don't think it's good practice to assume that strings are ASCII unless they're appropriately typed.

Reply:
The exercise spec promises ASCII, but maybe it should be changed, it's not terribly realistic, after all.

Reply:
I think that would  broaden the scope of the exercise too much. I just think we shouldn't encourage them to assume it's ASCII when implementing a generic method from the Julia code PoV, i.e. while it's, imo, fine to point out an implementation that works better for ASCII, they should also be aware to dispatch on `ASCIIString` (or w/e it's called nowadays) or add a check.

Reply:
There's no base ASCIIStr and the one in Strs.jl performs validation. I think this is an okay performance optimisation, if documented :shrug: 

I can remove this suggestion, though.